include/deal.II/base/smartpointer.h

```diff
@@ -396,6 +423,36 @@ SmartPointer<T, P>::operator=(const SmartPointer<T, P> &tt)
 
 
 
+template <typename T, typename P>
+inline SmartPointer<T, P> &
+SmartPointer<T, P>::operator=(SmartPointer<T, P> &&tt) noexcept
+{
+  // if objects on the left and right hand side of the operator= are
+  // the same, then this is a no-op
+  if (&tt != this)
+    {
+      // Let us unsubscribe from the current object
+      if (pointed_to_object_is_alive && t != nullptr)
+        t->unsubscribe(&pointed_to_object_is_alive, id);
+
+      // Then reset to the new object, and subscribe to it, assuming
+      // that the the rhs points to anything in the first place:
+      if (tt.pointed_to_object_is_alive && tt != nullptr)
```

Comment:
I think it would make more sense to subscribe to a dead object than to do nothing here - that way we would still get the correct error message when we try to access it later on.

Reply:
No no, you can't! You're going to call a member function on an object that no longer exists. This is bad. I should just assert that `tt` points to a live object here.

Reply:
You're right.

I have to wrap up a few non-deal.II things before the workshop - I'll get back to this on Saturday or Sunday.